File: csharp/Animal.cs
using System;
using System.Diagnostics;

abstract class Animal(string name)
{
    public string Name { get; } = name;

    public virtual string Speak() => $"{Name} says {Sound()}";
    
    public abstract string Sound();
}

class Cow(string name) : Animal(name) { public override string Sound() => "moooo"; }
class Horse(string name) : Animal(name) { public override string Sound() => "neigh"; }
class Sheep(string name) : Animal(name) { public override string Sound() => "baaaa"; }

static class AnimalPlay
{
    static void Main()
    {
        Animal h = new Horse("CJ");
        Debug.Assert(h.Speak() == "CJ says neigh");
        Cow c = new("Bessie");
        Debug.Assert(c.Speak() == "Bessie says moooo");
        Debug.Assert(new Sheep("Little Lamb").Speak() == "Little Lamb says baaaa");
    }
}

File: php/animal.php
<?php

abstract class Animal {
    public $name;
    public function __construct($name) {
        $this->name = $name;
    }
    public function speak() { 
        return "$this->name says {$this->sound()}"; 
    }
    abstract public function sound() : string;
}

class Cow extends Animal {
    public function sound() : string {
        return 'moooo';
    }
}

class Horse extends Animal {
    public function sound() : string {
        return 'neigh';
    }
}

class Sheep extends Animal {
    public function sound() : string {
        return 'baaaa';
    }
}

$h = new Horse('CJ');
assert($h->speak() == 'CJ says neigh');
$c = new Cow('Bessie');
assert($c->speak() == 'Bessie says moooo');
assert(new Sheep('Little Lamb')->speak() == 'Little Lamb says baaaa');

?>